src/main/java/nl/b3p/tailormap/api/solr/SolrHelper.java

```diff
@@ -0,0 +1,330 @@
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.request.schema.SchemaRequest;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.client.solrj.response.UpdateResponse;
+import org.apache.solr.client.solrj.response.schema.SchemaResponse;
+import org.apache.solr.common.SolrDocumentList;
+import org.apache.solr.common.SolrException;
+import org.geotools.api.data.Query;
+import org.geotools.api.data.SimpleFeatureSource;
+import org.geotools.api.feature.simple.SimpleFeature;
+import org.geotools.data.simple.SimpleFeatureCollection;
+import org.geotools.data.simple.SimpleFeatureIterator;
+import org.locationtech.jts.geom.Geometry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Solr utility/wrapper class. This class provides methods to add or update a full-text feature type
+ * index for a layer, find in the index for a layer, and clear the index for a layer. It also
+ * provides a method to close the Solr client as well as automatically closing the client when used
+ * in a try-with-resources.
+ */
+public class SolrHelper implements AutoCloseable, Constants {
+  private final SolrClient solrClient;
+  private final FeatureSourceFactoryHelper featureSourceFactoryHelper;
+
+  private static final Logger logger =
+      LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  public static final int SOLR_BATCH_SIZE = 1000;
+  // milliseconds
+  private static final int SOLR_TIMEOUT = 7000;
+
+  /**
+   * Constructor.
+   *
+   * @param solrClient the Solr client, this will be closed when this class is closed
+   * @param featureSourceFactoryHelper the feature source factory helper
+   */
+  public SolrHelper(
+      @NotNull SolrClient solrClient,
+      @NotNull FeatureSourceFactoryHelper featureSourceFactoryHelper) {
+    this.solrClient = solrClient;
+    this.featureSourceFactoryHelper = featureSourceFactoryHelper;
+  }
+
+  public static String getIndexLayerId(String serviceId, String layerName) {
+    return (serviceId + "__" + layerName).replaceAll(":", "__");
+  }
+
+  /**
+   * Add or update a feature type index for a layer.
+   *
+   * @param indexLayerId the layer name, as created by using {@link #getIndexLayerId(String,
+   *     String)}
+   * @param tmFeatureType the feature type
+   * @throws UnsupportedOperationException if the operation is not supported, possibly because not
+   *     search field shave been defined
+   * @throws IOException if an I/O error occurs
+   * @throws SolrServerException if a Solr error occurs
+   */
+  public void addFeatureTypeIndexForLayer(
+      @NotNull String indexLayerId, @NotNull TMFeatureType tmFeatureType)
+      throws UnsupportedOperationException, IOException, SolrServerException {
+
+    createSchemaIfNotExists();
+    if (null == tmFeatureType) {
```

Comment:
Due to maxFeatures returned by WFS'es paging (if supported) might be required to completely index a WFS feature type. For now maybe just throw an exception if trying to index a WFS feature source?